medmodels/data_synthesis/synthesizer.py

```diff
@@ -0,0 +1,132 @@
+"""Abstract class for synthesizers."""
+
+from __future__ import annotations
+
+from abc import ABCMeta, abstractmethod
+from pathlib import Path
+from typing import TYPE_CHECKING, Any, Literal, Optional, Type
+
+import torch
+from torch import nn
+
+from medmodels.data_synthesis.synthesizer_model import SynthesizerModel
+from medmodels.medrecord.medrecord import MedRecord
+
+if TYPE_CHECKING:
+    from medmodels.data_synthesis.builder import SynthesizerBuilder
+
+
+class Synthesizer(nn.Module, metaclass=ABCMeta):
+    """Synthesizer is the abstract class to define synthesizers."""
+
+    _preprocessor: Type[nn.Module]
+    _postprocessor: Type[nn.Module]
+    device: torch.device
+
+    def __init__(
+        self,
+        preprocessor: Type[nn.Module],
+        postprocessor: Type[nn.Module],
+        device: Literal["cpu", "cuda"] = "cuda",
+    ) -> None:
+        """Initializes the synthesizer with the required components.
+
+        Args:
+            preprocessor (Type[nn.Module]): Type of preprocessor module.
+            postprocessor (Type[nn.Module]): Type of postprocessor module.
+            device (Literal["cpu", "cuda"], optional): Device to use for the model. Defaults
+                to "cuda".
+        """
+        super(Synthesizer, self).__init__()
+        self._preprocessor = preprocessor
+        self._postprocessor = postprocessor
+        if device in ["cuda"] and not torch.cuda.is_available():
+            device = "cpu"
+        self.device = torch.device(device)
+        self.to(self.device)
+
+    @staticmethod
+    def builder() -> SynthesizerBuilder:
+        """Creates a SynthesizerBuilder instance.
+
+        Returns:
+            SynthesizerBuilder: An instance of SynthesizerBuilder.
+        """
+        from medmodels.data_synthesis.builder import SynthesizerBuilder
+
+        return SynthesizerBuilder()
+
+    @abstractmethod
+    def fit(
+        self,
+        medrecord: MedRecord,
+        checkpoint_directory: Optional[Path],
+        **kwargs: Any,
```

Comment:
if you really want to pass anything as kwargs then `Dict[str, Any]` was correct. I though that you once said that these kwargs were for extra paths or sth, then `Dict[str, Path]` but I probably remember wrong.

Can you give me an example of which additional kwargs might be used in the "child method"? Because from what i udnerstand now, this would violate the concept of subclasses. The entire point of ABC ist to define which methods exist, including their parameters (all of them, subclasses should not have different parameters as this defeats the purpose of using ABC)

Reply:
either all of these attributes are used in the abstract method definition (and therefore any future synthesizer implementations) or we need to rethink the whole ABC and subclass structure

Reply:
It is something very specific to MTGAN, do not know whether it makes sense to add it there. Although it could be added for now and then rethought if more synthesizers are subsequently added?

Reply:
I'd prefer if we chose the correct solution now tbh.

If the attributes of the actual implementations of the functions are not the same, they cannot be defined as abstract functions. That would defeat the purpose of ABC.
If this is the case, we should remove those abstract definitions.

Reply:
Then, removing all abstract methods that could not be shared? Or removing the abstract classes themselves?

Reply:
remove all abstract methods (where the actual implementations require different attributes) and if all abstract methods are removed, also remove the abstract declaration of the class (the `metaclass=ABCMeta` part)